src/transfers-history/client.ts

```diff
@@ -0,0 +1,132 @@
+
+export enum TransfersHistoryEvent {
+  TransfersUpdated = "TransfersUpdated",
+}
+
+export type TransfersUpdatedEventListenerParams = {
+  depositorAddr: string;
+  filledTransfersCount: number;
+  pendingTransfersCount: number;
+};
+
+export type TransfersUpdatedEventListener = (params: TransfersUpdatedEventListenerParams) => void;
+export type TransfersHistoryClientEventListener = TransfersUpdatedEventListener;
+export type TransfersHistoryClientParams = {
+  chains: {
+    chainId: ChainId;
+    providerUrl: string;
+  }[];
+  pollingIntervalSeconds?: number;
+};
+
+export class TransfersHistoryClient {
+  private eventEmitter = new EventEmitter();
+  private web3Providers: Record<ChainId, providers.Provider> = {};
+  private spokePoolInstances: Record<ChainId, SpokePool> = {};
+  private eventsQueriers: Record<ChainId, SpokePoolEventsQuerier> = {};
+  private eventsServices: Record<string, Record<ChainId, SpokePoolEventsQueryService>> = {};
+  private pollingIntervalSeconds: number = 15;
+  private pollingTimers: Record<string, NodeJS.Timer[]> = {};
+
+  constructor(
+    config: TransfersHistoryClientParams,
+    private logger = new Logger(),
+    private transfersRepository = new TransfersRepository()
+  ) {
+    if (config.pollingIntervalSeconds) {
+      this.pollingIntervalSeconds = config.pollingIntervalSeconds;
+    }
+
+    for (const chain of config.chains) {
+      clientConfig.web3ProvidersUrls[chain.chainId] = chain.providerUrl;
+      this.web3Providers[chain.chainId] = new providers.JsonRpcProvider(chain.providerUrl);
+      this.spokePoolInstances[chain.chainId] = SpokePool__factory.connect(
+        clientConfig.spokePools[chain.chainId].addr,
+        this.web3Providers[chain.chainId]
+      );
+      this.eventsQueriers[chain.chainId] = new SpokePoolEventsQuerier(
+        this.spokePoolInstances[chain.chainId],
+        undefined,
+        this.logger
+      );
+    }
+  }
+
+  public setLogLevel(level: LogLevel) {
+    this.logger.setLevel(level);
+  }
+
+  public async startFetchingTransfers(depositorAddr: string) {
+    const timers = this.pollingTimers[depositorAddr];
+    if (timers) throw new Error(`Address ${depositorAddr} is already monitored`)
+
+    this.initSpokePoolEventsQueryServices(depositorAddr);
+    this.getEventsForDepositor(depositorAddr);
+    const timer = setInterval(() => {
+      this.getEventsForDepositor(depositorAddr);
+    }, this.pollingIntervalSeconds * 1000);
+    this.pollingTimers[depositorAddr] = [...(this.pollingTimers[depositorAddr] || []), timer];
+  }
+
+  public stopFetchingTransfers(depositorAddr: string) {
+    const timers = this.pollingTimers[depositorAddr];
+
+    if (timers) {
+      timers.map(timer => clearInterval(timer));
+      delete this.pollingTimers[depositorAddr];
+    }
+  }
+
+  public on(event: TransfersHistoryEvent, cb: TransfersHistoryClientEventListener) {
+    this.eventEmitter.on(event, cb);
+  }
```

Comment:
If event emitters are the primary way to communicate state updates with the outside world, they may not play particularly nicely with react-query's async-await usage, as they may need to be wrapped in promises.

Reply:
They said react-query has relatively nice support for this kind of async updates